Fix isNumeric

kamil-tekiela opened the pull request:
Follow-up from #17758

I looked into it more. MySQL doesn't require quotes for integers, floats or decimals. phpMyAdmin already abstracts the MySQL types with two: TYPE_INT and TYPE_REAL. There's already a method called `isNumericType()` so why not simplify it all and get rid of `isNumeric` completely?

kamil-tekiela committed:
Fix isNumeric

Signed-off-by: Kamil Tekiela <tekiela246@gmail.com>

williamdes commented on test/classes/Controllers/Export/ExportControllerTest.php:
```diff
@@ -168,6 +168,6 @@ public function testExportController(): void
         $exportController($request);
         $output = $this->getActualOutputForAssertion();
-        $this->assertStringContainsString(htmlspecialchars($expectedOutput), $output);
+        $this->assertStringContainsString(htmlspecialchars($expectedOutput, ENT_QUOTES), $output);
```
only quotes ?
is this not displayed in a text area ?

kamil-tekiela replied:
Regardless of where it will be displayed, we should do a thorough review of all usages of `htmlspecialchars` and `htmlentities` in the whole codebase. PHP 8.1 changed the default value. Previously it was `ENT_COMPAT` and now it's `ENT_QUOTES | ENT_SUBSTITUTE | ENT_HTML401`. 

For this PR, I wanted to make the change as small as possible to ensure that the tests pass. As ENT_QUOTES can be considered a superset of ENT_COMPAT, that is it encodes everything ENT_COMPAT does and more, there should be no backwards incompatible changes. All it changes is that from now on single quotes will also be encoded.

williamdes replied:
Oh okay, its not great that `ENT_QUOTES` also encodes simple quotes `'`. But I understand the change. Why not `ENT_COMPAT` ?
here htmlentities is just some code to protect html properties I think

> All it changes is that from now on single quotes will also be encoded.

Yeah but I see not good reason to encode them

kamil-tekiela replied:
Makes sense. Changed.

williamdes commented:
> I looked into it more. MySQL doesn't require quotes for integers, floats or decimals. phpMyAdmin already abstracts the MySQL types with two: TYPE_INT and TYPE_REAL. There's already a method called isNumericType() so why not simplify it all and get rid of isNumeric completely?

Okay but be aware that in the export where `isNumericType` is used you may considerably change the performance.
Use a pre-evaluated property stored in a property. Calling getters costs a lot. That's why I created the `isXyz` properties

kamil-tekiela commented:
> Okay but be aware that in the export where isNumericType is used you may considerably change the performance.
Use a pre-evaluated property stored in a property. Calling getters costs a lot. That's why I created the isXyz properties

Thanks for the heads up. I will benchmark the change now.

williamdes commented:
> > Okay but be aware that in the export where isNumericType is used you may considerably change the performance.
> > Use a pre-evaluated property stored in a property. Calling getters costs a lot. That's why I created the isXyz properties
> 
> Thanks for the heads up. I will benchmark the change now.

Let me know :)

I used xhprof: https://github.com/phpmyadmin/phpmyadmin/wiki/How-to-use-xhprof
Having a working xhprof profile file viewer is a bit complicated but I can write a wiki article for it

kamil-tekiela commented:
@williamdes The performance is much worse with the additional function call. I reverted back to readonly property. But I removed the `isNumericType` method. I did it to reduce confusion as now we would have two methods that produce the same result. I kept the getter for `isNumeric` but it's not used anywhere except two tests. Is it acceptable?

williamdes reviewed (approved):
Did you check there is not regressions regarding #16896 ?

kamil-tekiela commented:
I cannot find any refressions.
